### PR TITLE
Added use_dpi setting to WindowDescriptor

### LINKED ISSUE #1118: The ability to disable the recently added hidpi window resizing feature
**What problem does this solve or what need does it fill?**

When the user of a game sets the "Windowed Resolution" to 1920x1080, they **always** mean physical pixels. Not being able to do this often results in rendering artifacts and definitely results in recording issues when using software like OBS.

Adjusting a Game to work with 4k resolutions is no different than getting a game to work between 800x600 and 1920x1080 resolutions. I myself, already had a Bevy project that I was recording/streaming the development of, that successfully scaled from 640x360 up to 4k and beyond.

Because I can't tell Bevy to use physical pixels when creating the Window, the new hidpi support is a breaking change that prevents developers and users from being able to control the created Window size.

**Describe the solution would you like?**

Add some sort of option to WindowDescriptor allowing me to specify whether I want to use Physical, or DPI scaled pixels for determining the window's size.

**Describe the alternative(s) you've considered?**

Alternatives include, allowing me to override the "scaling factor" so I can set it to 1.0, effectively disabling window scaling that way.

**Additional context**

My previous work in Bevy 0.2 having implemented resolution-aware rendering:

https://youtu.be/0vOHM2RZ508?t=6598

I have more recent records I could probably upload. But from that timestamp there's a pretty clear example of what I got going.

Here's a relevant article concerning another Game Engine that demonstrate the many pitfalls of scaling games; especially pixel art games:  https://alvarber.gitlab.io/pixel-perfection-in-godot.html

### PR DESCRIPTION
**Pull request by TheRawMeatball**
Maybe closes #1118

### CODE REVIEW

File: crates/bevy_winit/src/lib.rs [hunk 1]
```diff
@@ -68,13 +68,15 @@ fn change_window(_: &mut World, resources: &mut Resources) {
                     window.set_title(&title);
                 }
                 bevy_window::WindowCommand::SetResolution {
-                    resolution: (logical_width, logical_height),
+                    resolution: r,
```
Comment (DJMcNab) [inline crates/bevy_winit/src/lib.rs:71]:
```suggestion
                    resolution: (width, height),
```

File: crates/bevy_winit/src/lib.rs [hunk 2]
```diff
@@ -68,13 +68,15 @@ fn change_window(_: &mut World, resources: &mut Resources) {
...
                 } => {
                     let window = winit_windows.get_window(id).unwrap();
-                    window.set_inner_size(winit::dpi::LogicalSize::new(
-                        logical_width,
-                        logical_height,
-                    ));
+                    if logical {
+                        window.set_inner_size(winit::dpi::LogicalSize::new(r.0, r.1));
+                    } else {
+                        window.set_inner_size(winit::dpi::PhysicalSize::new(r.0, r.1));
```
Comment (DJMcNab) [inline crates/bevy_winit/src/lib.rs:78]:
```suggestion
                        window.set_inner_size(winit::dpi::LogicalSize::new(width, height));
                    } else {
                        window.set_inner_size(winit::dpi::PhysicalSize::new(width, height));
```

File: crates/bevy_ui/src/flex/mod.rs [hunk 3]
```diff
@@ -179,7 +179,11 @@ pub fn flex_node_system(
 
     // assume one window for time being...
     let logical_to_physical_factor = if let Some(primary_window) = windows.get_primary() {
-        primary_window.scale_factor()
```
Comment (DJMcNab) [inline crates/bevy_ui/src/flex/mod.rs:182]: Or just make `Window::scale_factor()` return `1` on a physical sized window?

File: crates/bevy_winit/src/winit_windows.rs [hunk 4]
```diff
@@ -38,13 +38,16 @@ impl WinitWindows {
...
-                    window_descriptor.height,
-                ))
-                .with_resizable(window_descriptor.resizable)
-                .with_decorations(window_descriptor.decorations),
+            _ => {
+                let s = (window_descriptor.width, window_descriptor.height);
+                if window_descriptor.use_dpi {
+                    winit_window_builder.with_inner_size(winit::dpi::LogicalSize::new(s.0, s.1))
+                } else {
+                    winit_window_builder.with_inner_size(winit::dpi::PhysicalSize::new(s.0, s.1))
```
Comment (DJMcNab) [inline crates/bevy_winit/src/winit_windows.rs:46]:
```suggestion
                let WindowDescriptor {width, height, ..} = window_descriptor;
                if window_descriptor.use_dpi {
                    winit_window_builder.with_inner_size(winit::dpi::LogicalSize::new(width, height))
                } else {
                    winit_window_builder.with_inner_size(winit::dpi::PhysicalSize::new(width, height))
```

File: examples/ui/ui.rs [hunk 5]
```diff
@@ -3,6 +3,9 @@ use bevy::prelude::*;
 /// This example illustrates the various features of Bevy UI.
 fn main() {
     App::build()
+        .add_resource(WindowDescriptor {
```
Comment (DJMcNab) [inline examples/ui/ui.rs:6]: Why is this needed?

**Review by DJMcNab — APPROVED** [review]
Overall looks good - there are a few changes which could be useful, but nothing obviously blocking.

File: examples/window/window_settings.rs [hunk 6]
```diff
@@ -7,6 +7,7 @@ fn main() {
             title: "I am a window!".to_string(),
             width: 500.,
             height: 300.,
+            use_dpi: false,
```
[the remark this answers is not in the document]
Comment (TheRawMeatball) [reply examples/window/window_settings.rs:10]: Sure, this was mostly for testing during development. I'll create a nicer example

File: crates/bevy_ui/src/flex/mod.rs (hunk 3 above)
Comment (TheRawMeatball) [reply crates/bevy_ui/src/flex/mod.rs:182]: I kept `Window::scale_factor()` as is so people who opt out of hidpi can still access the scale factor when implementing an alternative system.

Comment (DJMcNab) [reply crates/bevy_ui/src/flex/mod.rs:182]: That makes sense. I just noticed that you have already reimplemented that same logic three times, which isn't ideal. Maybe `monitor_scale_factor` vs `[window_]scale_factor` would be sensible? It's unclear what would be best.

File: examples/ui/ui.rs (hunk 5 above)
Comment (TheRawMeatball) [reply examples/ui/ui.rs:6]: It was for testing, forgot to remove it.

File: crates/bevy_ui/src/flex/mod.rs (hunk 3 above)
Comment (TheRawMeatball) [reply crates/bevy_ui/src/flex/mod.rs:182]: Diffrent windows can have diffrent scale factors, so I would choose window_scaling_factor.

**Comment by razaamir** [comment]
Can I be removed from this thread

**Comment by dlpatri** [comment]
Wow, I wasn't expecting such quick action on my request. This is awesome. Thanks guys.

**Comment by DJMcNab** [comment]
> Can I be removed from this thread

You need to click unsubscribe.

File: examples/window/window_settings.rs [hunk 7]
```diff
@@ -34,3 +36,18 @@ fn toggle_cursor(input: Res<Input<KeyCode>>, mut windows: ResMut<Windows>) {
         window.set_cursor_visibility(!window.cursor_visible());
     }
 }
+
+/// This system toggles DPI usage when enter is pressed
+fn toggle_dpi(input: Res<Input<KeyCode>>, mut windows: ResMut<Windows>) {
```
Comment (DJMcNab) [inline examples/window/window_settings.rs:41]: 🔥 
It would be quite good to have some text/ui which ignores the physical resolution in this example, or else add this code to a ui example.

**Comment by DJMcNab** [comment]
On my machine the new UI from your example is somewhat broken in higher scaling modes, which doesn't make sense. Also I feel like we should call some stuff `use_scaling` etc instead of use_dpi, but I'm not sure.

**Comment by TheRawMeatball** [comment]
Can you post some details and a screenshot?

**Comment by DJMcNab** [comment]
So in theory, because the UI is all done based on logical pixels, the images should be identical (barring scaling artifacts). However, 100% gives me:
image
and 125% gives me
image
where for some reason the left panel stays the same width, and the font doesn't move down.

**Comment by DJMcNab** [comment]
That is definitely better, although there's still some jank - if you go through this video frame by frame for example you can see a few examples.

https://user-images.githubusercontent.com/36049421/102925674-c28b8b00-448b-11eb-8327-ac6d0f5be1ef.mp4

That being said, it's not super important, because we don't really expect people to be changing window scaling on/off that often ™️

**Comment by TheRawMeatball** [comment]
Yeah, I'm fine with it stuttering when its being cycled at such a frequency. Given that it requires a full UI redraw on every switch, its not very surprising.

**Comment by blunted2night** [comment]
A couple of thoughts. Would `system_scale_factor` be a better name that `true_scale_factor`? Would it be simpler to make `scale_factor` the current scale factor instead of having to check `use_dpi` every where? This would either be the user selected value, or automatically updated when the system changes. Does it make sense to include the scale factor as part of the window resized event since these are tightly coupled?

**Comment by TheRawMeatball** [comment]
> Would it be simpler to make scale_factor the current scale factor instead of having to check use_dpi every where? This would either be the user selected value, or automatically updated when the system changes.

I don't exactly understand what you're offering here. Do you mean you want to change `Window.scale_factor` so it stores 1 when using physical pixels? In such a case, we'd have to store the scaling factor from winit in a seperate variable, effectively caching the result of `scale_factor()`. I dont think its worth the additional complexity.  
> Does it make sense to include the scale factor as part of the window resized event since these are tightly coupled?

They aren't tightly coupled. Both are triggered by seperate events, and are handled by diffrent pieces of code.

**Comment by DJMcNab** [comment]
I think they mean that when scale_factor is ignored, the scale_factor method should return `1.`

i.e. the current scale_factor method should be 
```rust
pub fn scale_factor(&self)->f32{
    if self.use_dpi { 1 } else{ self.scale_factor }
}
```
and there should be a
```rust
pub fn requested_scale_factor(&self)->f32 {self.scale_factor}
```

**Comment by TheRawMeatball** [comment]
This is the current behaviour, thats why I am confused.

**Comment by blunted2night** [comment]
> 
> 
> > Would it be simpler to make scale_factor the current scale factor instead of having to check use_dpi every where? This would either be the user selected value, or automatically updated when the system changes.
> 
> I don't exactly understand what you're offering here. Do you mean you want to change `Window.scale_factor` so it stores 1 when using physical pixels? In such a case, we'd have to store the scaling factor from winit in a seperate variable, effectively caching the result of `scale_factor()`. I dont think its worth the additional complexity.

Yeah, I guess it might be nice to specify a custom scale factor in addition to disabling the system value. And the scale_factor variable would always hold the current value, with another holding the system scale factor if required. While subjective, to me having `scale_factor()` perform logic to know what the real current scale factor is more complex than a separate variable recording the current system scale_factor independent of the currently selected scale factor.

> 
> > Does it make sense to include the scale factor as part of the window resized event since these are tightly coupled?
> 
> They aren't tightly coupled. Both are triggered by seperate events, and are handled by diffrent pieces of code.

I was thinking that changes to the scaling factor will effect the resolution reported and from the users perspective it might be better to receive both of these at the same time.

**Comment by dlpatri** [comment]
Being able to supply your own "scale_factor" could be useful for testing.

**Comment by TheRawMeatball** [comment]
I made a major rework to implement @blunted2night s suggestions and clean everything up.

File: crates/bevy_winit/src/lib.rs [hunk 8]
```diff
@@ -332,27 +331,36 @@ pub fn winit_runner(mut app: App) {
...
                     let mut windows = app.resources.get_mut::<Windows>().unwrap();
                     let window_id = winit_windows.get_window_id(winit_window_id).unwrap();
                     let window = windows.get_mut(window_id).unwrap();
-                    if window.use_dpi() {
+                    if let Some(sf) = window.scale_factor_override() {
+                        *new_inner_size = winit::dpi::LogicalSize::new(
+                            window.physical_width(),
+                            window.physical_height(),
+                        )
+                        .to_physical(sf)
```
Comment (blunted2night) [inline crates/bevy_winit/src/lib.rs:346]: This block doesn't appear to have any effect, if I am reading it right, it updates the `new_inner_size` with  a value that will never be read.

Comment (blunted2night) [reply crates/bevy_winit/src/lib.rs:346]: I imagine it was intended to recalculate the logical size and update the back-end. If so I believe the calculation would require going from the windows logical size to the pysical size via the reported scale factor, then from the physical size to the user logical size via the scale factor override.

Comment (TheRawMeatball) [reply crates/bevy_winit/src/lib.rs:346]: This value will be read by winit to change the window size. [Reference](https://docs.rs/winit/0.24.0/winit/event/enum.WindowEvent.html#variant.ScaleFactorChanged)

Comment (blunted2night) [reply crates/bevy_winit/src/lib.rs:346]: Sorry, the inner_size is already in physical units. On window at least, the window can be resize when crossing into a display with a different scaling factor and resolution, so some action should be taken to update the window and issue notifications.

Comment (blunted2night) [reply crates/bevy_winit/src/lib.rs:347]: OK my bad, I am still confused by constructing a logical size from a the current physical size, then converting that to physical size. Wouldn't that cause the window size to jump every time a scale factor change is received?

Shouldn't it be something like:

```rust
*new_inner_size = winit::dpi::PhysicalSize::new(
  window.physical_width(),
  window.physical_height(),
);
```

Comment (TheRawMeatball) [reply crates/bevy_winit/src/lib.rs:347]: The current system does something quite similar. keep in mind that `sf` is the override, and will not change when the true scale factor changes. Creating a logical size and converting it to physical is just a neater way to write this:
```rust
*new_inner_size = winit::dpi::PhysicalSize::new(
  window.physical_width() * sf,
  window.physical_height() * sf,
);
```
(Probably wrong due to types)

Comment (blunted2night) [reply crates/bevy_winit/src/lib.rs:347]: But, unless I am missing something, you are multiplying the current physical height by the logical to physical height scaling factor and setting that as the new physical size. If the scaling factor override is 1.0 it is a nop. If it is something else, then whenever the user changes their system scaling factor, the window will increase in size by the overridden scaling factor.

Comment (TheRawMeatball) [reply crates/bevy_winit/src/lib.rs]: Yup, that was a bug, and now its gone.

**Comment by cart** [comment]
rebased on master

**Comment by cart** [comment]
I made a few tweaks. Feel free to contest any of them:

* Removed window.set_physical_resolution(). its a nice shorthand, but i think it over-complicates things. Too much implicit behavior when it comes state changes (could change scale factor _or_ requested resolution). Id prefer to just let people override scale factor to 1.0 then use set_resolution
* Moved the scale factor override example logic to its own example. I want to keep the window_settings example simple
* Renamed  "true scale factor" cases to "backend scale factor". "true" felt a bit too arbitrary / undefined to me. 
* Renamed instances of DPI to "scale factor". DPI != scale factor. DPI is a measure of pixel density whereas scale factor is a scale preference.
* simplified event handling logic 
* style tweaks

**Comment by TheRawMeatball** [comment]
Sounds good!
